doc/compiler/jitserver/Messaging.md

```diff
@@ -0,0 +1,162 @@
+<!--
+Copyright (c) 2018, 2021 IBM Corp. and others
+
+This program and the accompanying materials are made available under
+the terms of the Eclipse Public License 2.0 which accompanies this
+distribution and is available at https://www.eclipse.org/legal/epl-2.0/
+or the Apache License, Version 2.0 which accompanies this distribution and
+is available at https://www.apache.org/licenses/LICENSE-2.0.
+
+This Source Code may also be made available under the following
+Secondary Licenses when the conditions for such availability set
+forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+General Public License, version 2 with the GNU Classpath
+Exception [1] and GNU General Public License, version 2 with the
+OpenJDK Assembly Exception [2].
+
+[1] https://www.gnu.org/software/classpath/license.html
+[2] http://openjdk.java.net/legal/assembly-exception.html
+
+SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+-->
+
+# JITServer Messaging Protocol
+
+## Overview
+
+All data exchanged between server and clients is structured in the form of "JITServer Messsages". Every message sent requires a response, i.e. messages are exchanged in pairs. Both client and server can send messages. Client sends only compilation requests, while server sends all other message types.
+
+Every JITServer message (outgoing and incoming) has the following properties:
+
+- It has a `MessageType`, as defined in `runtime/compiler/net/MessageTypes.hpp`
+- It contains one or more serializable datapoints. Serializable datapoints are defined
+as either trivially copyable datatypes, or as datatypes with `onRecv()` and `onSend()` functions specialized in `RawTypeConvert` (see [Supported Datatypes](#supported-datatypes)).
+- The message type defines the structure of a message, i.e. the type determines the number, order and types of datapoints. For every message, we need to send some specific data, and receive a specific response.
+
+> **Note:** Since at least one datapoint is required, it's not technically possible to send an empty message. To achieve the same result, send `JITServer::Void` and then read it on the receiving end.
+
+Messages are not usually constructed directly. Instead, `Client/ServerStream::write()` and `Client/ServerStream::read()` methods are used to send a message as a variadic argument list, and read it as a tuple (See the [example](#example) below).
+
+For the messaging protocol to work correctly, client and server must run matching JITServer versions. If a new message type gets added, or the structure of an existing message is modified, the version number will increment.
```

Comment:
'the version number will increment' --> 'the protocol version number needs to be incremented'